File: 0703_Kth_Largest_Element_in_a_Stream/solution.java
package cn.huang.leetcode;

import java.util.PriorityQueue;

/*
    703. Kth Largest Element in a Stream
    Design a class to find the kth largest element in a stream. Note that it is the kth largest element in the sorted order,
    not the kth distinct element.

    Your KthLargest class will have a constructor which accepts an integer k and an integer array nums, which contains
    initial elements from the stream. For each call to the method KthLargest.add, return the element representing the kth
    largest element in the stream.
 */
public class LC_0703_KthLargestElementinaStream {
    class KthLargest {
        private PriorityQueue<Integer> q;
        private int size;
        public KthLargest(int k, int[] nums) {
            //优先队列的作用是能保证每次取出的元素都是队列中权值最小的
            q = new PriorityQueue<>();
            size = k;

            for (int i=0;i<nums.length;i++){
                add(nums[i]);
            }

        }

        public int add(int val) {
            if (q.size() < size) {
                q.add(val);
            } else if (q.peek() < val) {
                //获取并删除队首元素
                q.poll();
                q.add(val);
            }
            return q.peek();
        }
    }
}

File: 0703_Kth_Largest_Element_in_a_Stream/solution.cpp
/*
703. Kth Largest Element in a Stream

Level: Easy

https://leetcode.com/problems/kth-largest-element-in-a-stream
*/
#include <vector>
#include <string>
#include <unordered_map>
#include <unordered_set>

#include <algorithm>
#include <queue>

#include "common/ListNode.h"
#include "common/TreeNode.h"

using namespace std;

/*
  Solution: 
*/

class KthLargest
{
private:
  //return the largest (by default use less<int>) element
  priority_queue<int, vector<int>, greater<int>> q;
  int size;

public:
  KthLargest(int k, vector<int> &nums)
  {
    size = k;

    for (int num : nums)
    {
      add(num);
    }
  }

  int add(int val)
  {
    if (q.size() < size)
    {
      q.push(val);
    }
    else if (q.top() < val)
    {
      //获取并删除队首元素
      q.pop();
      q.push(val);
    }
    return q.top();
  }
};

/**
 * Your KthLargest object will be instantiated and called as such:
 * KthLargest* obj = new KthLargest(k, nums);
 * int param_1 = obj->add(val);
 */